### ISSUE TITLE
`LoadBalancerExchangeFilterFunction` throws an exception if, on startup, the `LBEFF` can't fid any instances.

### ISSUE THREAD

**Issue by joshlong** [spring-cloud/spring-cloud-commons#386]
If i write this code:

```@LoadBalanced @Bean WebClient wc (WebClient.Builder b ){  return b.build();  } ```

it works if I start the node that uses the `WebClient` *after* I've started the service and the service registry. If, howevver, I start the client and the client can't loadbalance because there are none in the registry, then i get this error:

```
1.080  INFO [reservation-client,76f49b74dea0f7ac,76f49b74dea0f7ac,true] 28309 --- [server-epoll-10] c.n.u.concurrent.ShutdownEnabledTimer    : Shutdown hook installed for: NFLoadBalancer-PingTimer-reservation-service
2018-07-01 17:32:11.098  INFO [reservation-client,76f49b74dea0f7ac,76f49b74dea0f7ac,true] 28309 --- [server-epoll-10] c.netflix.loadbalancer.BaseLoadBalancer  : Client: reservation-service instantiated a LoadBalancer: DynamicServerListLoadBalancer:{NFLoadBalancer:name=reservation-service,current list of Servers=[],Load balancer stats=Zone stats: {},Server stats: []}ServerList:null
2018-07-01 17:32:11.101  INFO [reservation-client,76f49b74dea0f7ac,76f49b74dea0f7ac,true] 28309 --- [server-epoll-10] c.n.l.DynamicServerListLoadBalancer      : Using serverListUpdater PollingServerListUpdater
2018-07-01 17:32:11.104  INFO [reservation-client,76f49b74dea0f7ac,76f49b74dea0f7ac,true] 28309 --- [server-epoll-10] c.n.l.DynamicServerListLoadBalancer      : DynamicServerListLoadBalancer for client reservation-service initialized: DynamicServerListLoadBalancer:{NFLoadBalancer:name=reservation-service,current list of Servers=[],Load balancer stats=Zone stats: {},Server stats: []}ServerList:org.springframework.cloud.netflix.ribbon.eureka.DomainExtractingServerList@38da4995
2018-07-01 17:32:11.166 ERROR [reservation-client,76f49b74dea0f7ac,76f49b74dea0f7ac,true] 28309 --- [server-epoll-10] .a.w.r.e.DefaultErrorWebExceptionHandler : Failed to handle request [GET http://localhost:9999/reservations/names]

java.lang.RuntimeException: java.lang.IllegalArgumentException: instance can not be null
	at com.example.reservationclient.ReservationClientApplication.lambda$webClient$0(ReservationClientApplication.java:68) ~[classes/:na]
	at org.springframework.web.reactive.function.client.ExchangeFilterFunction.lambda$andThen$1(ExchangeFilterFunction.java:56) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.web.reactive.function.client.ExchangeFilterFunction.lambda$apply$2(ExchangeFilterFunction.java:67) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.web.reactive.function.client.DefaultWebClient$DefaultRequestBodyUriSpec.exchange(DefaultWebClient.java:317) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.web.reactive.function.client.DefaultWebClient$DefaultRequestBodyUriSpec.retrieve(DefaultWebClient.java:364) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at com.example.reservationclient.ReservationClientApplication.lambda$routes$1(ReservationClientApplication.java:84) ~[classes/:na]
	at org.springframework.web.reactive.function.server.support.HandlerFunctionAdapter.handle(HandlerFunctionAdapter.java:61) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.web.reactive.DispatcherHandler.invokeHandler(DispatcherHandler.java:168) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at org.springframework.web.reactive.DispatcherHandler.lambda$handle$1(DispatcherHandler.java:160) ~[spring-webflux-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118) [reactor-core-3.1.8.RELEASE.jar:3.1.8.RELEASE]
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.onNext(ScopePassingSpanSubscriber.java:81) [spring-cloud-sleuth-core-2.0.0.RELEASE.jar:2.0.0.RELEASE]
	at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:67) ~[reactor-core-3.1.8.RELEASE.jar:3.1.8.RELEASE]
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.onNext(ScopePassingSpanSubscriber.java:81) [spring-cloud-sleuth-core-2.0.0.RELEASE.jar:2.0.0.RELEASE]
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:76) ~[reactor-core-3.1.8.RELEASE.jar:3.1.8.RELEASE]
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.onNext(ScopePassingSpanSubscriber.java:81) [spring-cloud-sleuth-core-2.0.0.RELEASE.jar:2.0.0.RELEASE]
	at reactor.core.publisher.FluxConcatMap$ConcatMap
```

This exception is thrown in the filter itself, before an HTTP request is ever issued, and as such it's not in the `WebClient` itself. This means that my normal recourse for handling errors in the web request, like `doOnErrorResume ( Exception.class, ex -> ...) ` don't apply. 

It doesn't make a difference if i use `HystrixCommands`, either.

So, i end up having to do: 

```	@Bean
		WebClient webClient(LoadBalancerExchangeFilterFunction eff) {
				ExchangeFilterFunction fallback = (clientRequest, exchangeFunction) -> {
						try {
								return eff.filter(clientRequest, exchangeFunction);
						}
						catch (Exception e) {
								return Mono.just(ClientResponse.create(HttpStatus.SERVICE_UNAVAILABLE).build());
						}
				};
				return WebClient.builder()
					.filter(fallback)
					.build();
		}
```

when ceating the `WebClient`. This works. My Hystrix circuit breaker ALWAYS returns the fallback value. Please consider doing something out of the bx like this in the `LoadBalancerExchangeFilterFunction` s that i don't have to wrap with this try/catch.

**Comment by MustaphaCh** [comment]
@joshlong I have the same problem I 'am trying to setup the workaround that you propose, but I don't find the method create in ClientResponse, I use the release 5.0.7.

**Comment by jruhe-adesso** [comment]
I face the same issue. It is a clear violation of the reactive paradigm, that I have to try-catch

`        Flux<AbStub> abStubs = abServiceAdapter.findReferencing(singletonList(bcStub))
                .replay()
                .autoConnect();`

where

`public Flux<AbStub> findReferencing(Collection<dStub> dStubs) {
        return this.webClient.get()
                .uri(findReferencingUri, dStubs.stream()
                        .map(dStub -> dStub.getProductId())
                        .collect(Collectors.joining(",")))
                .retrieve()
                .onStatus(HttpStatus::isError, clientResponse -> Mono.just(new IllegalStateException("Unexpected response by ab service: " + clientResponse.statusCode())))
                .bodyToFlux(AbStub.class);
    }`

retrieve() throws an exception, when LoadBalancerExchangeFilterFunction cannot resolve an instance.

No IO resources should be accessed prior any subscription to the flux returned by WebClient.

**Comment by ryanjbaxter** [comment]
Closed via 5248f3f
